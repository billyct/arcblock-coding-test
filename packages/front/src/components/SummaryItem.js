import {useState} from 'react'
import PropTypes from 'prop-types'
import cn from 'classnames'

const SummaryItem = ({title, tip, children}) => {

  const [show, setShow] = useState(false)

  return (
    <div className='flex justify-between items-start text-sm leading-loose px-6 py-2 hover:bg-gray-200'>

      <div
        data-testid='SummaryItem__title'
        className='text-gray-600 w-24 flex-shrink-0 relative'
        onMouseEnter={() => setShow(!!tip)}
        onMouseLeave={() => setShow(false)}
      >
        <span>
          {title}
        </span>

        <div
          data-testid='SummaryItem__tip'
          className={cn('px-2 py-1 bg-black rounded text-white left-0 absolute whitespace-nowrap z-10', {
            'hidden': !show,
            'block': show,
          })}
        >
          {tip}
        </div>
      </div>
      <div className='break-all'>{children}</div>
    </div>
  )
}

SummaryItem.propTypes = {
  title: PropTypes.string.isRequired,
  tip: PropTypes.string,
}

SummaryItem.defaultProps = {
  tip: '',
}

export default SummaryItem
