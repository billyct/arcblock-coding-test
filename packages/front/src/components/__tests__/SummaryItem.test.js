import {fireEvent, render} from '@testing-library/react'
import SummaryItem from '../SummaryItem'

describe('test SummaryItem Component', () => {
  test('snapshot should correctly', () => {
    const {container} = render(
      <SummaryItem title='a title' tip='a tip'>
        a children
      </SummaryItem>
    )

    expect(container).toMatchSnapshot()
  })

  test(`tip's visibility should correctly`, () => {
    const {getByTestId} = render(
      <SummaryItem title='Hash' tip='a tip'/>
    )

    const title = getByTestId('SummaryItem__title')
    const tip = getByTestId('SummaryItem__tip')

    fireEvent.mouseEnter(title)
    expect(tip).toHaveClass('block')

    fireEvent.mouseLeave(title)
    expect(tip).toHaveClass('hidden')
  })

  test(`tip should hidden when tip is empty`, () => {
    const {getByTestId} = render(
      <SummaryItem title='Hash'/>
    )

    const title = getByTestId('SummaryItem__title')
    const tip = getByTestId('SummaryItem__tip')

    fireEvent.mouseEnter(title)
    expect(tip).not.toHaveClass('block')
  })
})
