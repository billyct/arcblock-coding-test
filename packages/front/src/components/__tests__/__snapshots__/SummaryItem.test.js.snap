// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SummaryItem Component snapshot should correctly 1`] = `
<div>
  <div
    class="flex justify-between items-start text-sm leading-loose px-6 py-2 hover:bg-gray-200"
  >
    <div
      class="text-gray-600 w-24 flex-shrink-0 relative"
      data-testid="SummaryItem__title"
    >
      <span>
        a title
      </span>
      <div
        class="px-2 py-1 bg-black rounded text-white left-0 absolute whitespace-nowrap z-10 hidden"
        data-testid="SummaryItem__tip"
      >
        a tip
      </div>
    </div>
    <div
      class="break-all"
    >
      a children
    </div>
  </div>
</div>
`;
